sql/core/src/test/scala/org/apache/spark/sql/test/SharedSQLContext.scala

```diff
@@ -17,4 +17,37 @@
 
 package org.apache.spark.sql.test
 
-trait SharedSQLContext extends SQLTestUtils with SharedSparkSession
+trait SharedSQLContext extends SQLTestUtils with SharedSparkSession {
+
+  /**
+   * Suites extending [[SharedSQLContext]] are sharing resources (eg. SparkSession) in their tests.
+   * Such resources are initialized by the suite before thread audit takes thread snapshot and
```

Comment:
Sorry but this still does not explain why this is happening. It's just stating that it is.

For example, in `SharedSparkSession`, there is this code:

```
  protected override def afterAll(): Unit = {
    super.afterAll()
    if (_spark != null) {
      _spark.sessionState.catalog.reset()
      _spark.stop()
      _spark = null
    }
  }
```

If you move `super.afterAll()` to after the session is stopped, won't that solve the problem and avoid this?

Reply:
Your suggestion solves one part of the problem. The other one lies here:

```
  protected override def beforeAll(): Unit = {
    initializeSession()

    // Ensure we have initialized the context before calling parent code
    super.beforeAll()
  }
```

Session initialized before thread snapshot. This should also happen in the opposite order. Because I've seen the comment in the code I decided not to change it.

Reply:
Why is the latter a problem? At worst you'll have less thread after the suite finishes than when it started, which should be fine, no? The problem is having leaked threads, not the other way around.

Reply:
Ok, I think I see your point. Still, the comment here is confusing. Can't this be done in `SharedSparkSession` instead, where that initialization happens, so that it's clear what it's talking about?

Reply:
I think this is an easier to understand comment about what's going on here:

```
  /**
   * Suites extending [[SharedSQLContext]] are sharing resources (eg. SparkSession) in their tests.
   * Trat trait initializes the spark session in its [[beforeAll()]] implementation before the
   * automatic thread snapshot is performed, so the audit code could fail to report threads leaked
   * by that shared session.
   *
   * The behavior is overridden here to take the snapshot before the spark session is initialized.
   */
```

Sorry for the noise.

Reply:
Much better phrased and compressed explanation, applied. Agree that it would be better to move this functionality into `SharedSparkSession` on the other hand it would lead far in terms of number of modifications. `SharedSparkSession` has to extend has to extend `SparkFunSuite` which I don't see it worth the effort. The other option what I see also doesn't help in terms of understanding. Namely moving manual thread audit into `SparkFunSuite` and leaving `enableAutoThreadAudit = false` in `SharedSQLContext` but splitting functionality rarely can help. Ideas?